src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp

```diff
@@ -372,6 +384,89 @@ void runtime::gpu::CudaKernelBuilder::get_replace_slice_op(
+    //  product = tid
+    //  d0 = product/stride[0]
+    //  product = product % stride[0]
+    //  d1 = product/stride[1]
+    //  ...
+    writer << "int coordinate_product = " << i_thread_index << ";\n";
+    for (size_t i = 0; i < rank; i++)
+    {
+        writer << "int " << o_coordinates << i << " = division_by_invariant_multiplication("
+               << "coordinate_product, " << i_stride_magic << "[" << i << "], " << i_stride_shift
+               << "[" << i << "]);\n";
+        writer << "coordinate_product -= (" << o_coordinates << i << " * " << i_strides << "[" << i
+               << "]);\n";
+    }
+
+    // index into reduced tensor from coordinates of non-reduced tensor
+    std::string reduced_idx = "reduced_idx";
+    writer << "int " << reduced_idx << " = 0;\n";
+    for (size_t i = 0; i < rank; i++)
+    {
+        writer << "reduced_idx += " << o_coordinates << i << " * " << i_reduced_strides << "[" << i
+               << "];\n";
+    }
+
+    return reduced_idx;
+}
+
+void runtime::gpu::CudaKernelBuilder::get_softmax_op(codegen::CodeWriter& writer,
+                                                     const std::string& name,
+                                                     const std::array<std::string, 2>& data_types,
+                                                     const size_t rank)
+{
+    writer << "extern \"C\" __global__ void cuda_" << name << "(" << data_types[0] << "* in, "
+           << data_types[1] << "* out, "
+           << "int* strides, "
+           << "int* stride_magic, "
+           << "int* stride_shift, "
+           << "int* reduced_strides, "
+           << "float alpha, float beta, "
+           << "size_t nthreads"
+           << ")\n";
+    writer.block_begin();
+    {
+        writer << "const int tid = blockDim.x*blockIdx.x + threadIdx.x;\n";
+        writer << "if (tid < nthreads)\n";
+        writer.block_begin();
+        {
+            // calculate tensor coordinates
+            std::string reduced_idx = reduce_coordinate_transform_helper(writer,
+                                                                         "tid",
+                                                                         "strides",
+                                                                         "stride_magic",
+                                                                         "stride_shift",
+                                                                         "reduced_strides",
+                                                                         "coordinate",
+                                                                         rank);
+
+            // TODO: mediate atomic memory access contention
+            writer << data_types[0] << " val = expf(load(in, tid));\n";
+            writer << "atomicAdd(&out[" << reduced_idx << "], val);\n";
+            writer << "__threadfence();\n";
+            writer << data_types[1] << " sum = out[" << reduced_idx << "];\n";
```

Comment:
This will only work when all data of reduced axis fall in the same block.

Reply:
Softmax is now implemented via fused elementwise collective operations. See gpu_emitter.cpp.